### DISCUSSION TITLE
Getting error: can't make objects of type android.hardware.camera2.CameraDevice$StateCallback while integration of Nativescript-8 with camera2 android api

### DISCUSSION THREAD

**Discussion by kssinghal980** [NativeScript/NativeScript#10132] [Q&A]
Trying to call openCamera method of CameraManager of android in Nativescript version 8, getting the error while trying to create an instance of StateCallback.

Code for reference:
```ts
return new class extends android.hardware.camera2.CameraDevice.StateCallback {
            constructor(context) {
                super()
                return global.__native(context)
            }
            onOpened(camera: any) {
                console.log("camera on opened hit: ", !!camera);
            }
            onDisconnected(camera: any) {
                console.log("camera on disconnected hit");
            }
            onError(camera: any, error: any) {
                console.log("error occured while opening camera",(error));
            }
            onClosed(param: android.hardware.camera2.CameraDevice) {
                CLog("onClosed in statecallback ", param);
            }
        }(this);
```
Device error log:
```
JNI DETECTED ERROR IN APPLICATION: can't make objects of type android.hardware.camera2.CameraDevice$StateCallback: 0x71809820
 in call to NewObjectA
```
NativeScript version: 8.3.3

**Comment by farfromrefug** [comment]
@kssinghal980 You need to add the `@NativeClass` decorator for this to work with N 8.x https://docs.nativescript.org/best-practices/native-class.html

**Reply by kssinghal980** [reply]
Thank you for your response. 

After adding @NativeClass, getting below error now:

ReferenceError: NativeClass is not defined.

tsconfig.json file for reference:

`{
    "compilerOptions": {
        "target": "es2017",
        "module": "esnext",
        "declaration": true,
        "removeComments": false,
        "noLib": false,
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "skipLibCheck": true,
        "lib": [
            "es6",
            "dom",
            "es2017"
        ],
        "sourceMap": true,
        "pretty": true,
        "allowUnreachableCode": false,
        "allowUnusedLabels": false,
        "noEmitHelpers": true,
        "noEmitOnError": false,
        "noImplicitAny": false,
        "noImplicitReturns": true,
        "noImplicitUseStrict": false,
        "noFallthroughCasesInSwitch": true,
        "moduleResolution": "node"
    },
    "exclude": [
        "node_modules"
    ],
    "compileOnSave": false
}
`

**Reply by farfromrefug** [reply]
@kssinghal980 is that your app or a plugin? 
You need a ts transformer to do that. If in an app the transformer is added but tsc still need to be patched.
Explained here https://blog.nativescript.org/nativescript-7-for-plugin-authors/

**Reply by kssinghal980** [reply]
I have created a custom camera plugin and while using that plugin to any sample app, getting the error

**Reply by kefahB** [reply]
@kssinghal980 as @farfromrefug pointed, you need `ts-patch` in order to get @NativeClass working

tsconfig.json
```ts
"compilerOptions": {
  "plugins": [{
    "transform": "@nativescript/webpack/transformers/ns-transform-native-classes",
    "type": "raw"
  }]
}
```

Package.json
```ts
"scripts": {
  "setup": "npm i && ts-patch install"
}
```
